[Root7] A path to reducing global object registration and management.

hageboeck opened the pull request:
This is the start of a longer-term effort to transform ROOT into working with less globals, making ownership of objects clearer, and going more towards standard C++ behaviour instead of implicit ownership. One could call this the "ROOT 7 mode". In order to allow for early testing, this mode is introduced as an opt-in feature in ROOT 6.

## What this PR brings

`TROOT.h` gets three new functions (names up for discussion):
```c++
   namespace ROOT::Experimental {
   void EnableObjectAutoRegistration();
   void DisableObjectAutoRegistration();
   bool ObjectAutoRegistrationEnabled();
   }
```
https://github.com/hageboeck/root/blob/040980428a842a3314c75bde1ba9e6ae208fe8b1/core/base/inc/TROOT.h#L100-L104

The naming and API is inspired by `EnableImplicitMT()` and its colleagues. These determine whether TObject-derived classes auto-register themselves to `gDirectory` when created. A well-known example is TH1 and all its derived classes, but what's covered by these functions will be extended further on the path to ROOT 7. Currently, this mode also covers RooFit objects, TGraph, and TEfficiency.
An extensive list of objects with auto registration and their state (maybe not fully complete) can be found below.

A lot of preparatory work was done to prepare ROOT to work both with auto registration on and off. A detailed list of this work can be found at the end. Currently, there is one commit that switches auto registration off across the entire CI to demonstrate that ROOT already today works in this mode. Once the PR is ready to merge, this commit will be removed to not break the workflows of ROOT's users who did not opt in. A single (or a few; TBD) builds with auto registration off will be added to the CI to keep test coverage for this feature.

### Difference to `TH1::AddDirectoryStatus()`
The ROOT 7 mode with auto registration off might look similar to `TH1::AddDirectoryStatus()`, but there are important differences:
- DirectoryStatus is a static (global) bool while `ObjectAutoRegistrationEnabled()` is `thread_local`
- DirectoryStatus only affects histogram-like objects (see table at the end). AutoRegistrationEnabled should eventually extend to more cases, e.g. RooFit objects, TEventList (planned), functions (planned).
- DirectoryStatus affects the behaviour of `TFile::Get`. `file->Get("name"); file->Get("name")` returns the same object when directory status is `true`, and it reads a new object every time when it's `false`. I consider it best to not interfere here, because one can use TFile if such behaviour is desired, and RFile if not. `AutoRegistrationEnabled()` should only decide what happens to new objects, not objects read from a file.
- DirectoryStatus cannot be set from the environment, and its default cannot be changed. It is a non-thread-safe runtime feature. `AutoRegistrationEnabled()` can be initialised from the environment (thread safe), and its state can be changed per thread, which might be vital for using it in MT frameworks.

### Setting defaults 
Some users or frameworks might be interested in operating in ROOT 7 mode already today, but they don't want to add calls to the above functions in each thread of their code. Especially for testing, they should be able to set a default without recompiling. Defaults can be set in a .rootrc:
```
Root.ObjectAutoRegistration: 0
```
and also using an environment variable:
```
ROOT_OBJECT_AUTO_REGISTRATION=0 ./executable --args
```

When any object queries `ObjectAutoRegistrationEnabled()`, a thread-safe initialisation is performed, taking defaults from the environment above. If no defaults are set, registration defaults to "on" in ROOT 6 and "off" in ROOT 7.

This, however, might lead to an "it worked on my machine" problem: Some users might have altered their environment, while others might have chosen the opposite. To make it obvious when a ROOT 6 runs in ROOT 7 preview mode, info messages such as the following are printed if the environment modifies the default:
```shell
$ ./executable --args
Info in <TROOT>: Object auto registration disabled using the environment variable ROOT_OBJECT_AUTO_REGISTRATION
```
(A similar message exists for rootrc)


## List of classes covered (or intended to be covered) by "DisableObjectAutoRegistration"

|                                           | Auto registration disabled with `DisableObjectAutoRegistration()` | How could it be disabled before?      | Comments                                                                                |
| --------------------------------------------- | --------------------------------- | ----------------------------- | ----------------------------------------------------------------------------------------------------- |
| TH1 and derived                                           | Yes                                 | TH1::AddDirectoryStatus()     |                                                                                                       |
| TGraph2D                                      | Yes                                 | TH1::AddDirectoryStatus()     |                                                                                                       |
| RooPlot                                       | Yes                                 | RooPlot::addDirectoryStatus() |                                                                                                       |
| TEfficiency                                   | Yes (New)                           | Was inconsistent     |                                                                                                       |
| TProfile2D                         | Yes                                 | TH1::AddDirectoryStatus()             |                                                                                                       |
| TEntryList                                    | Yes (in the future)                                 |  Not possible                             | To be done later                                                                                                     |
| TEventList                                    | Yes (in the future)                                 |  Not possible                             |  To be done later                                                                                                     |
| TTree                                         | Probably no (to be decided)                                 |                               | Unclear if it should be changed. It would be broken without being attached to a file                           |
| TBranch                                       | No                                 |                               | It needs to know what its tree is                                                   |
| TChain, TNTuple                               | No                                 |                               | Same as for TTree                                                                                    |
| TFunction | Probably yes, in the future. To be decided  |    | This is a registration not to `gDirectory` but to a global list of functions. Way forward to be decided. | 

### TODOs:
- [x] Add tests for implicit / explicit registration of TH1s
- [x] Deprecate `TH1::AddDirectory()` as suggested in #21523 ?
  - In Doxygen, it is already marked as deprecated, but a C++ deprecation should probably happen a bit later on the way to ROOT 7.
- [x] Merge the PRs below to make ROOT ready for running both with and without auto registration.
- [x] Remove the commit that globally switches auto registration off in the CI. Instead, create a dedicated build for this feature.

### Preparatory work
The following PRs were part of this branch, and have been split off over time to separate the introduction of this feature from preparing ROOT to work both with auto registration off and on:
- #20818
- #20830
- #20832
- #20833 
- #20886
- #20888 
- #20889
- #20919
- #20920
- #20921
- #20922 
- #20978
- #21224
- #21225
- #21226
- #21228
- #21229
- #21231
- #21248 
- #21287
- #21296 
- #21319
- #21341
- #21523
- #21587 
- #21646
- #21648 
- #21679
- #21708

hageboeck committed:
[core] Add ROOT::Experimental::DisableObjectAutoRegistration() & co.

- Add Enable/DisableObjectAutoRegistration() in ROOT::Experimental.
- Add a function to test if registration is on or off.
- Allow for setting defaults using
  - The environment variable ROOT_OBJECT_AUTO_REGISTRATION
  - The rootrc variable Root.ObjectAutoRegistration

pcanal commented on hist/hist/src/TH1.cxx:
```diff
@@ -741,7 +741,7 @@ TH1::TH1(const char *name,const char *title,Int_t nbins,const Double_t *xbins)
 Bool_t TH1::AddDirectoryStatus()
 {
-   return fgAddDirectory;
+   return !ROOT::ROOT7MemoryManagement() && fgAddDirectory;
```
Note that this flag is primarily about file constructions (and secondarily about shared ownership).  Beside the (intentional) memory leak it introduces in existing scripts, it leads most of them to not longer write the histograms to the file (because they are no longer attached to the File).  It is not clear whether this kind of approach is worth it compared to a more user active change (ie. Switching explicitly from TFile to RFile).

hageboeck committed:
[RF] Propagate implicit object registration settings to RooPlot.

hageboeck committed:
[PyROOT] Make a TFile context manager test/tutorial ready for ROOT 7.

- Don't rely on implicit registration to directories.
- Make the test insensitive to any defaults by always adding the
  histogram to the file.
- Make ruff happier.

hageboeck committed:
[NFC] Clarify the role of TH1::DirectoryAutoAdd.

The ROOT 7 ownership settings don't affect TH1::AddDirectoryStatus, so
objects read from a TFile associate themselves with that file. In light
of adding the ROOT 7 mode "auto registration off", the doxygen was
improved a bit.

hageboeck committed:
[RFile] Prepare an RFile test for both states of object auto registration.

hageboeck committed:
[hist] Add a test for explicit registration of TH1 to TDirectory.

hageboeck committed:
[roottest] Whitelist ROOT 7 preview messages in roottest checks.

- When the object auto registration is switched to ROOT 7 mode, the
corresponding notification needs to be masked from output diffs and/or
the RootTestDriver checks.
- Ruff formatting

hageboeck committed:
[hist] Propagate object registration settings to TH1/TEfficiency/TGraph2D.

In locations where TH1::AddDirectoryStatus is queried, also query
ObjectAutoRegistrationEnabled(). This extends the ROOT 7 testing mode to
TH1 and derived classes, TGraph2D and TEfficiency.

hageboeck committed:
[core] Add a test for thread-local object auto registration state.

pcanal commented on .github/workflows/root-ci.yml:
```diff
@@ -157,10 +157,11 @@ jobs:
           [ -d "${VIRTUAL_ENV_DIR}" ] && source ${VIRTUAL_ENV_DIR}/bin/activate
           echo "Python is now $(which python3) $(python3 --version)"
-          src/.github/workflows/root-ci-config/build_root.py  \
+          src/.github/workflows/root-ci-config/build_root.py \
```
```suggestion
          src/.github/workflows/root-ci-config/build_root.py  \
```

hageboeck replied:
This commit will be removed before merging. It is here to demonstrate that all tests work today also when object auto registration is off.

pcanal commented on core/base/src/TROOT.cxx:
```diff
@@ -293,6 +295,78 @@ namespace {
    }
+
+   enum class AutoReg : unsigned char {
+      NotInitialised = 0,
+      On,
+      Off,
```
```suggestion
      kNotInitialised = 0,
      kOn,
      kOff,
```

pcanal commented on core/base/src/TROOT.cxx:
```diff
@@ -293,6 +295,78 @@ namespace {
+         tlsState = defaultState;
+      }
+
+      return tlsState;
```
~Where is it documented that the run-time turning on and off is only a thread-local operation?~  Found it.

pcanal commented on core/base/src/TROOT.cxx:
```diff
@@ -293,6 +295,78 @@ namespace {
+      constexpr auto rcName = "Root.ObjectAutoRegistration";    // Update the docs if this is changed
+      constexpr auto envName = "ROOT_OBJECT_AUTO_REGISTRATION"; // Update the docs if this is changed
+
+      if (defaultState.load(std::memory_order_relaxed) == AutoReg::NotInitialised) {
```
Did you consider (instead of the atomic):
```suggestion
      static const AutoReg defaultState = []() {
      ....
      } ();
```

hageboeck replied:
This should work, thanks for the idea!

pcanal commented on core/base/src/TROOT.cxx:
```diff
@@ -616,6 +689,68 @@ namespace Internal {
+   ///
+   /// \note For classes deriving from TH1, both ObjectAutoRegistrationEnabled() and TH1::AddDirectoryStatus()
+   /// need to be true for auto-registration to take effect. The former should be preferred over the latter, because
+   /// it is thread safe and extends to more objects such as TGraph2D, TEfficiency, RooPlot.
```
```suggestion
   /// it is thread local and extends to more objects such as TGraph2D, TEfficiency, RooPlot.
```

dpiparo commented on core/base/src/TROOT.cxx:
```diff
@@ -616,6 +686,68 @@ namespace Internal {
+   /// The API might change without notice until moved out of this namespace.
+   namespace Experimental {
+   ////////////////////////////////////////////////////////////////////////////////
+   /// \brief Enable automatic registration of objects for the current thread (ROOT 6 default).
```
Sorry to be picky here: does it make sense to have a table of the classes the instances of which are affected by the switch in this documentation?
That table can be also updated when new classes are added, such as`TE*List`.

hageboeck replied:
Yes, the table appeared now after the latest rebase. It's, however, not in the function you marked (which cannot be called outside of `TROOT.cxx`), it's in the function `ROOT::Experimental::DisableObjectAutoRegistration()` and its siblings.

Once merged, the docs should look roughly like this:
<img width="1417" height="1056" alt="image" src="https://github.com/user-attachments/assets/9ca76ad2-7729-4bd1-be3a-0d5270ec3acd" />

dpiparo reviewed (approved):
Thanks for this PR and the massive preparatory work. I am proud to see ROOT managed to put some order in this very core functionality, to modernise itself and offer an even more rewarding user-experience.

hageboeck committed:
Add release notes for the auto registration off mode.

hageboeck committed:
[CI] Handle all linux environment settings in one step.

hageboeck committed:
[CI] Disable object auto registration in the clang ninja job.
